### ISSUE TITLE
v2 beta: Parsing of Link header fails because Mirakl uses commas to delimit query params

### ISSUE THREAD

**Issue by christophersjchow** [stripe/stripe-mirakl-connector#55]
Mirakl expects `,` to delimit commercial order ids when querying its OR11 API. The `getNextLink` function does an `explode` using `,` as a delimiter which incorrectly expands out the Link header.

Here are some examples of Link headers:

With prev and next:
```
<https://scentregroupau-dev.mirakl.net/api/orders?commercial_ids=3693596968,0195242688,3884560115,2845696461,5526974611,3962904573,0055278822,0253203047,8674454819,2173887289,9577383944,1730096837,7195291116,7569009629,4878630488,5259284619,7978839735,3766272697,9557235094,2201264206,1131931008,7338035900&max=2&offset=0>; rel="previous", <https://scentregroupau-dev.mirakl.net/api/orders?commercial_ids=3693596968,0195242688,3884560115,2845696461,5526974611,3962904573,0055278822,0253203047,8674454819,2173887289,9577383944,1730096837,7195291116,7569009629,4878630488,5259284619,7978839735,3766272697,9557235094,2201264206,1131931008,7338035900&max=2&offset=4>; rel="next"';
```

Just with next
```
<https://scentregroupau-dev.mirakl.net/api/orders?commercial_ids=3693596968,0195242688,3884560115,2845696461,5526974611,3962904573,0055278822,0253203047,8674454819,2173887289,9577383944,1730096837,7195291116,7569009629,4878630488,5259284619,7978839735,3766272697,9557235094,2201264206,1131931008,7338035900&max=2&offset=2>; rel="next";
```

Here is a potential patch though I think this function should be made `public` and unit tested.

```patch
diff --git a/src/Service/MiraklClient.php b/src/Service/MiraklClient.php
index 932c292..9f7d9d0 100644
--- a/src/Service/MiraklClient.php
+++ b/src/Service/MiraklClient.php
@@ -88,11 +88,14 @@ class MiraklClient
 
     private function getNextLink(ResponseInterface $response)
     {
-        static $pattern = '/<([^>]+)>;\s*rel="([^"]+)"/';
+        static $split_pattern = '/,\s+/';
+        static $link_pattern = '/<([^>]+)>;\s*rel="([^"]+)"/';
 
         $links = $response->getHeaders()['link'][0] ?? '';
-        foreach (explode(',', $links) as $link) {
-            if (1 === preg_match($pattern, trim($link), $res) && 'next' === $res[2]) {
+        $split_links = preg_split($split_pattern, $links);
+
+        foreach ($split_links as $link) {
+            if (1 === preg_match($link_pattern, trim($link), $res) && 'next' === $res[2]) {
                 return $res[1];
             }
         }
```

**Comment by christophersjchow** [comment]
I'm confused why the pagination doesn't seem to be working in this case. Here are the logs:

```
[2021-01-27 23:04:49] http_client.INFO: Request: "GET https://scentregroupau-dev.mirakl.net/api/payment/debit?max=10" [] []
[2021-01-27 23:04:49] http_client.INFO: Response: "200 https://scentregroupau-dev.mirakl.net/api/payment/debit?max=10" [] []
No mirakl orders pending debit
[2021-01-27 23:04:49] http_client.INFO: Request: "GET https://scentregroupau-dev.mirakl.net/api/orders?max=10&commercial_ids=3693596968%2C0195242688%2C3884560115%2C2845696461%2C5526974611%2C3962904573%2C0055278822%2C0253203047%2C8674454819%2C2173887289%2C9577383944%2C1730096837%2C7195291116%2C7569009629%2C4878630488%2C5259284619%2C7978839735%2C3766272697%2C9557235094%2C2201264206%2C1131931008%2C7338035900" [] []
[2021-01-27 23:04:50] http_client.INFO: Response: "200 https://scentregroupau-dev.mirakl.net/api/orders?max=10&commercial_ids=3693596968%2C0195242688%2C3884560115%2C2845696461%2C5526974611%2C3962904573%2C0055278822%2C0253203047%2C8674454819%2C2173887289%2C9577383944%2C1730096837%2C7195291116%2C7569009629%2C4878630488%2C5259284619%2C7978839735%2C3766272697%2C9557235094%2C2201264206%2C1131931008%2C7338035900" [] []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-accepted logistical order. {"commercial_id":3693596968,"order_id":"3693596968-A"} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-accepted logistical order. {"commercial_id":"0195242688","order_id":"0195242688-A"} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-accepted logistical order. {"commercial_id":3884560115,"order_id":"3884560115-A"} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-accepted logistical order. {"commercial_id":2845696461,"order_id":"2845696461-A"} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-accepted logistical order. {"commercial_id":5526974611,"order_id":"5526974611-B"} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-accepted logistical order. {"commercial_id":3962904573,"order_id":"3962904573-A"} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-accepted logistical order. {"commercial_id":"0055278822","order_id":"0055278822-A"} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-accepted logistical order. {"commercial_id":"0253203047","order_id":"0253203047-A"} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":8674454819} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":2173887289} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":9577383944} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":1730096837} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":7195291116} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":7569009629} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":4878630488} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":5259284619} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":7978839735} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":3766272697} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":9557235094} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":2201264206} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":1131931008} []
[2021-01-27 23:04:50] app.INFO: Skipping payment capture for non-existing commercial order. {"commercial_id":7338035900} []
```

However, `7338035900` definitely exists as I've verified by making the same call and increasing the `max`.

**Comment by christophersjchow** [comment]
Ok it looks to be a bug in `getNextLink`. This is something I might be able to fix.

**Comment by christophersjchow** [comment]
So the `Link` header that gets returned from Mirakl is: `<https://scentregroupau-dev.mirakl.net/api/orders?commercial_ids=3693596968,0195242688,3884560115,2845696461,5526974611,3962904573,0055278822,0253203047,8674454819,2173887289,9577383944,1730096837,7195291116,7569009629,4878630488,5259284619,7978839735,3766272697,9557235094,2201264206,1131931008,7338035900&max=10&offset=10>; rel="next"`

When I run the following test code:

```php
<?php

$pattern = '/<([^>]+)>;\s*rel="([^"]+)"/';
$links = '<https://scentregroupau-dev.mirakl.net/api/orders?commercial_ids=3693596968,0195242688,3884560115,2845696461,5526974611,3962904573,0055278822,0253203047,8674454819,2173887289,9577383944,1730096837,7195291116,7569009629,4878630488,5259284619,7978839735,3766272697,9557235094,2201264206,1131931008,7338035900&max=10&offset=10>; rel="next"';
$link = explode(',', $links);

print_r($link);
?>
```

The `explode` call incorrectly explodes out the query params:

```
Array
(
    [0] => <https://scentregroupau-dev.mirakl.net/api/orders?commercial_ids=3693596968
    [1] => 0195242688
    [2] => 3884560115
    [3] => 2845696461
    [4] => 5526974611
    [5] => 3962904573
    [6] => 0055278822
    [7] => 0253203047
    [8] => 8674454819
    [9] => 2173887289
    [10] => 9577383944
    [11] => 1730096837
    [12] => 7195291116
    [13] => 7569009629
    [14] => 4878630488
    [15] => 5259284619
    [16] => 7978839735
    [17] => 3766272697
    [18] => 9557235094
    [19] => 2201264206
    [20] => 1131931008
    [21] => 7338035900&max=10&offset=10>; rel="next"
)
```

As far as I can tel this can be fixed simply by applying this patch:

```patch
diff --git a/src/Service/MiraklClient.php b/src/Service/MiraklClient.php
index 932c292..fc43d17 100644
--- a/src/Service/MiraklClient.php
+++ b/src/Service/MiraklClient.php
@@ -91,7 +91,7 @@ class MiraklClient
         static $pattern = '/<([^>]+)>;\s*rel="([^"]+)"/';

         $links = $response->getHeaders()['link'][0] ?? '';
-        foreach (explode(',', $links) as $link) {
+        foreach ($links as $link) {
             if (1 === preg_match($pattern, trim($link), $res) && 'next' === $res[2]) {
                 return $res[1];
             }
```

**Comment by christophersjchow** [comment]
Ah I see its when there's a previous page we need to split.

**Comment by christophersjchow** [comment]
I can confirm the above patch does work for us.

**Comment by youssef-stripe** [comment]
Fixed by #58 and e9d9c73. Thanks for flagging and for the fix!
